controlplane/kubeadm/api/v1alpha4/kubeadmcontrolplanetemplate_webhook.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const kubeadmControlPlaneTemplateImmutableMsg = "KubeadmControlPlaneTemplate spec.template.spec field is immutable. Please create new resource instead."
+
+func (r *KubeadmControlPlaneTemplate) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-controlplane-cluster-x-k8s-io-v1alpha4-kubeadmcontrolplanetemplate,mutating=true,failurePolicy=fail,groups=controlplane.cluster.x-k8s.io,resources=kubeadmcontrolplanetemplates,versions=v1alpha4,name=default.kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+var _ webhook.Defaulter = &KubeadmControlPlaneTemplate{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (r *KubeadmControlPlaneTemplate) Default() {
+	defaultKubeadmControlPlaneSpec(&r.Spec.Template.Spec, r.Namespace)
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-controlplane-cluster-x-k8s-io-v1alpha4-kubeadmcontrolplanetemplate,mutating=false,failurePolicy=fail,groups=controlplane.cluster.x-k8s.io,resources=kubeadmcontrolplanetemplates,versions=v1alpha4,name=validation.kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1;v1beta1
+
+var _ webhook.Validator = &KubeadmControlPlaneTemplate{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *KubeadmControlPlaneTemplate) ValidateCreate() error {
+	spec := r.Spec.Template.Spec
+	allErrs := valdiateKubeadmControlPlaneSpec(spec, r.Namespace)
+	allErrs = append(allErrs, validateEtcd(&spec, nil)...)
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(GroupVersion.WithKind("KubeadmControlPlaneTemplate").GroupKind(), r.Name, allErrs)
+	}
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *KubeadmControlPlaneTemplate) ValidateUpdate(oldRaw runtime.Object) error {
+	var allErrs field.ErrorList
+	old := oldRaw.(*KubeadmControlPlaneTemplate)
```

Comment:
We usually seem to check if the type case works and then return an error (instead of a panic which would happen otherwise), e.g.:
```bash
        oldM, ok := old.(*Machine)
	if !ok {
		return apierrors.NewBadRequest(fmt.Sprintf("expected a Machine but got a %T", old))
	}
```

Reply:
Can you also fix it in the KubeadmControlPlane webhook? (that one was already there before your PR :))

Reply:
I added the checks in both the places now :)  
I dont think the check was present before (or atleast I [didnt find it](https://github.com/kubernetes-sigs/cluster-api/blob/master/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_webhook.go#L146) 😅 ).
Anyway, it should be taken care now.

Reply:
I was referring to that one: https://github.com/kubernetes-sigs/cluster-api/blob/44ed423ee6857fbd0751d4ca8b822aa9c8fa4d36/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_webhook.go#L150